apm-agent-api/src/main/java/co/elastic/apm/api/ElasticApm.java

```diff
@@ -260,24 +260,42 @@ public static void setServiceInfoForClassLoader(@Nullable ClassLoader classLoade
      * and calling getConfig("transaction_sample_rate") would provide the current value
      * </p>
      * <p>
-     * Note the value returned can be null if an invalid key is used, and also
-     * if the "public-api" set of instrumentations has not been applied
+     * If an invalid key is passed as the first argument, or if the incorrect type for
+     * the key is passed as the second argument, the method will throw an IllegalArgumentException
+     * </p>
+     * <p>
+     * Note the value returned can be null if the "public-api" set of instrumentations
+     * has not been applied and also if the agent hasn't finished initializing
      * </p>
      *
      * @param key the string option
+     * @param type the type of the option (String, Long, Integer, etc)
      * @return The current value for the option, per the option type (String, Long, Integer, etc)
      * @since 1.37.0
      */
-    public static Object getConfig(String key) {
-        // co.elastic.apm.api.ElasticApmInstrumentation.ConfigInstrumentation.getConfig
+    public static <T> T getConfig(String key, Class<T> type) {
+        Object value = doGetConfig(key);
+        if (value == null) {
+            throw new IllegalArgumentException("There is no such option: " + key);
+        } else if (value instanceof IllegalStateException) {
+            return null;
+        } else if (value.getClass().equals(type)) {
+            return (T) value;
+        } else {
+            throw new IllegalArgumentException("The option: '"+key+"' is not of type "+type.getName());
+        }
```

Comment:
[minor] I think we could simplify this further by keeping the exception creation in the advice code, then only check if the returned value is an exception (and if yes, then throw it), also the type matching is not exact, and we could for example pass `Number` to get a `Long` configuration option, but that's definitely nit-picking here :-).

Reply:
it's similar branching just in the private method than the advice - I prefer keeping the advice as simple as possible.

I did think about looking for inheritance, but decided to keep it exact matching because I think that way the customer reads the option more thoroughly - and we seem to have a lot of scanning the options rather than reading them. But I'm not stuck on this opinion

Reply:
actually I decided you are right and changed it to an inheritance match